### PR TITLE
Uid from ldap

### LINKED ISSUE #33: Populate uid data model
Requires #32.

Uid values should be obtained during user account creation when a user logs in via SSO. We therefore need to extend the `create_user` method of `imperial_coldfront_plugin.oidc.ICLOIDCAuthenticationBackend` to do this.

The source of the UID data is an ldap server available at `unixldap.cc.ic.ac.uk`. Further details on this to follow.

The logic of the `create_user` method should be:
- Check if ldap server uri is configured in settings.
- If yes:
  - Query the ldap server using the username accessed from `claims`.
  - If successful:
    - get uid value from ldap server response
    - create user
    - create uid model instance linked to user
  - else raise an error
- If no then create user

### PR DESCRIPTION
**Pull request by cc-a**
# Description

This pull request introduces LDAP integration to the allow retrieving and storing unix uid values for users.

User uid values are retrieved during the user creation process by the OIDC authentication backend provided by `imperial_coldfront_plugin.oidc.ICLOIDCAuthenticationBackend`. If the necessary settings for the LDAP server are present then the user's uid is retrieved from `unixldap.cc.ic.ac.uk`. If the uid cannot be retrieved then user creation is aborted. If the uid is successfully retrieved then a `UnixUID` instance is created to store the uid.

The uid retrieval process is tied to the OIDC authentication method as an underlying assumption is that the user must correspond to valid Imperial user in order for data to exist on the LDAP server.

The LDAP package used adds some additional OS requirements so a corresponding PR has been created for the development environment - imperialcollegelondon/coldfront_development_environment#2. You must use the branch from that PR to test this code.

Another requirement for trying this code out is that the OIDC authentication mechanism must be used. This is supported by the development environment but requires the correct environment variables to be set. I suggest we have a dedicated session to set these up for everybody. You can still create and login with local users (added via command line or the Django admin interface) but these will not have an associated UID value. Later on this will impose some limitations on what can be done with those users but `UnixUID` values can always be added manually.

To test:
- checkout the `ldap-support` branch of the development environment.
- checkout this branch in the submodule.
- make sure you have the appropriate environment variables set for OIDC - if done you will see "Log in via OpenID Connect" as an option.
- make sure you are on the College VPN or ZScaler
- start the app and log in using OIDC
- open a shell `docker compose exec app coldfront shell` and check a UnixUID instance was created for your user, e.g.:
```python
from imperial_coldfront_plugin.models import UnixUID
UnixUID.objects.get(user__username="yourusername")
```
on success a UnixUID will be successfully retrieved.

Fixes #33 

## Type of change

- [ ] Documentation (non-breaking change that adds or improves the documentation)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Optimization (non-breaking, back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (whatever its nature)

## Key checklist

- [ ] All tests pass (eg. `python -m pytest`)
- [ ] The documentation builds and looks OK (eg. `python -m sphinx -b html docs docs/build`)
- [ ] Pre-commit hooks run successfully (eg. `pre-commit run --all-files`)

## Further checks

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added or an issue has been opened to tackle that in the future. (Indicate issue here: # (issue))

### CODE REVIEW

**Review by AdrianDAlessandro — COMMENTED** [review]
I haven't fully tested that this works, but I have one suggestion with the settings, and a question:

How is this different from the [existing LDAP User Search coldfront plugin](https://coldfront.readthedocs.io/en/latest/config/#ldap-user-search) and/or the [LDAP Auth plugin](https://coldfront.readthedocs.io/en/latest/config/#ldap-auth)

**Comment by SaranjeetKaur** [comment]
Not yet tested. But a quick question @cc-a. You mentioned that 0 is a special value (refers to the root user or superadmin). when uids are concerned. Are there any other such special values/numbers or any other integer values would are acceptable? Negatives also?

**Comment by cc-a** [comment]
@SaranjeetKaur yes. I suspect you're thinking we should do some more validation on the database model and you're correct that would be a good idea. The picture is a little complex - https://en.wikipedia.org/wiki/User_identifier#Reserved_ranges - in short there are some common conventions but not any universal rules (except possibly for 0). To do a good job on this we'd need to find out from the identity team which conventions (if any) they follow when generating uids for users. We do know that negative values should not be allowed however.

**Comment by SaranjeetKaur** [comment]
The other thing I tried just now was logging into the app.
So I ran `docker compose exec app coldfront createsuperuser` and intially it said  `service "app" is not running` then I started  docker and it seemed to work. It asked for the following input from me:

```
Username (leave blank to use 'nobody'):          
Email address:
Password: 
Password (again):
```

And finally it said `Superuser created successfully.`. 
So it created me as a Superuser - does that mean that my uid is 0? Anybody who runs that command will have their uid as 0? Which means these uids are not unique? Is that ok?

**Comment by SaranjeetKaur** [comment]
> @SaranjeetKaur Not to worry, just a case of overlapping terminology. A superuser in the Django app effectively has all permissions within the webapp but that's disticnt from the `uid` value which uniquely identitfies a user when they access RCS infrastructure e.g. the HPC cluster.

Ah I see!!

File: pyproject.toml [hunk 1]
```diff
@@ -92,4 +93,4 @@ pydocstyle.convention = "google"
 ] # Missing docstring in public module, Missing docstring in public package
 
 [tool.django-stubs]
-django_settings_module = "imperial_coldfront_plugin" # a lie but seems to satisfy
+django_settings_module = "imperial_coldfront_plugin.settings"
```
[the remark this answers is not in the document]
Comment (cc-a) [reply pyproject.toml:96]: Anecdotally I can as you get mypy errors without this value so it must be having some effect.

File: imperial_coldfront_plugin/settings.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+"""Plugin settings that are imported into the global settings namespace of Coldfront."""
+
+from coldfront.config.env import ENV
+
+LDAP_SERVER_URI = ENV.str("LDAP_SERVER_URI", default="")
+LDAP_SEARCH_BASE = ENV.str("LDAP_SEARCH_BASE", default="")
```
[the remark this answers is not in the document]
Comment (cc-a) [reply imperial_coldfront_plugin/settings.py:6]: The implications of that pattern aren't totally clear to me. It relies on us also accessing all settings in the plugin via the local settings module. That has the implicit effect of adding the plugin settings into the global settings config. I think I probably prefer the explicit import from as that makes it clear in the project where and when the plugin settings are being included. I think we'd still also need to add the plugin to installed_apps on the project side so we'd still need something there.

**Comment by cc-a** [comment]
Closing this as I want to look at an alternative implementation.

**Pull request closed without merging**